### PR TITLE
Async rollrus

### PR DESCRIPTION
**Pull request by tamccall**
Made rollbar async improved performance from:

```
 2017-11-18 21:15:15 ⌚  andymc-MS-7808 in ~/go/src/github.com/benjamindow/rollrus
± |before ?:2 ✗| → go test -bench=. ./...
goos: linux
goarch: amd64
pkg: github.com/benjamindow/rollrus
BenchmarkVanillaLogger-4   	 1000000	      2758 ns/op
BenchmarkRollrusLogger-4   	      20	  77503141 ns/op
PASS
ok  	github.com/benjamindow/rollrus	4.889s
```

Edit: removing invalid benchmark

### CODE REVIEW

**Review by greggjs — APPROVED** [review]
I think these changes are awesome man. I wonder if `heroku/rollrus` would want these change?

**Comment by tamccall** [comment]
I think that the rollrus hook is getting closed and therefore not sending rollbar metrics. which is why i was able to get such a big perf gain

**Comment by tamccall** [comment]
Newest benchmark on pooled hook:

```
 2017-11-19 15:16:02 ⌚  andymc-MS-7808 in ~/go/src/github.com/benjamindow/rollrus
± |async-rollrus S:3 U:4 ?:2 ✗| → go test -bench=. ./... 
PASS
ok  	github.com/benjamindow/rollrus	0.003s
goos: linux
goarch: amd64
pkg: github.com/benjamindow/rollrus/benchmark
BenchmarkVanillaLogger-4   	 1000000	      2828 ns/op
BenchmarkRollrusLogger-4   	     500	  19151717 ns/op
PASS
ok  	github.com/benjamindow/rollrus/benchmark	12.806s
```

**Comment by tamccall** [comment]
```
 2017-11-19 21:16:56 ⌚  andymc-MS-7808 in ~/go/src/github.com/benjamindow/rollrus
± |async-rollrus {1} ?:2 ✗| → go test -bench=. ./... 
PASS
ok  	github.com/benjamindow/rollrus	0.002s
goos: linux
goarch: amd64
pkg: github.com/benjamindow/rollrus/benchmark
BenchmarkVanillaLogger-4               	 1000000	      2346 ns/op
BenchmarkRollrusLogger-4               	   10000	   1170229 ns/op
BenchmarkDownSamplingRollrusLogger-4   	  200000	      6409 ns/op
PASS
ok  	github.com/benjamindow/rollrus/benchmark	16.478s
?   	github.com/benjamindow/rollrus/throttle	[no test files]
```

**Comment by tamccall** [comment]
```
± |async-rollrus {2} U:2 ?:2 ✗| → go test -bench=. -count 3 ./... 
goos: linux
goarch: amd64
pkg: github.com/benjamindow/rollrus
BenchmarkVanillaLogger-4               	 1000000	      3145 ns/op
BenchmarkVanillaLogger-4               	  500000	      2970 ns/op
BenchmarkVanillaLogger-4               	  500000	      2995 ns/op
BenchmarkRollrusLogger-4               	    1000	   1809528 ns/op
BenchmarkRollrusLogger-4               	    2000	   1935645 ns/op
BenchmarkRollrusLogger-4               	    1000	   1848914 ns/op
BenchmarkDownSamplingRollrusLogger-4   	  200000	      6232 ns/op
BenchmarkDownSamplingRollrusLogger-4   	  200000	      6022 ns/op
BenchmarkDownSamplingRollrusLogger-4   	  200000	      6664 ns/op

```

File: rollrus.go [hunk 1]
```diff
@@ -2,60 +2,128 @@ package rollrus
...
+var defaultNumWorkers = 8 * runtime.NumCPU()
+var defaultBufferSize = 2 * defaultNumWorkers
+
 // Hook wrapper for the rollbar Client
 // May be used as a rollbar client itself
 type Hook struct {
 	roll.Client
-	triggers []log.Level
+	triggers      []log.Level
+	buffer        chan *log.Entry
```
Comment (tamccall) [inline rollrus.go]: Have considered replacing this with a diode interested to see what people thing

https://github.com/cloudfoundry/go-diodes

**Comment by tamccall** [comment]
No crazy throttling stuff new benchmarks:

```
± |async-rollrus-2 {2} U:2 ?:2 ✗| → go test -bench=. ./... 
goos: linux
goarch: amd64
pkg: github.com/benjamindow/rollrus
BenchmarkVanillaLogger-4       	 1000000	      1848 ns/op
BenchmarkWithChannelBuffer-4   	    1000	   2447418 ns/op
BenchmarkWithDiodeBuffer-4     	 1000000	      2350 ns/op
PASS
ok  	github.com/benjamindow/rollrus	8.855s
?   	github.com/benjamindow/rollrus/buffer	[no test files]
?   	github.com/benjamindow/rollrus/buffer/channel	[no test files]
?   	github.com/benjamindow/rollrus/buffer/diode	[no test files]
```

Need to add unit tests to the new buffers

File: rollrus.go [hunk 2]
```diff
@@ -2,60 +2,120 @@ package rollrus
...
-	return NewHookForLevels(token, env, defaultTriggerLevels)
+	return NewHookForLevels(token, env, RollrusConfig{})
 }
 
 // Setup a new hook with specified reporting levels, useful for adding to
 // your own logger instance.
-func NewHookForLevels(token string, env string, levels []log.Level) *Hook {
-	return &Hook{
+func NewHookForLevels(token string, env string, config RollrusConfig) *Hook {
+	if len(config.LogLevels) == 0 {
```
Comment (tamccall) [inline rollrus.go]: Might consider moving all of this into the config struct as an unexported method

File: benchmark_test.go [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+}
+
+func BenchmarkWithChannelBuffer(b *testing.B) {
+	if token == "" {
+		b.Skip("Could not get rollbar token")
+	}
+
+	rollrusLogger := logrus.New()
+	rollrusLogger.Out = ioutil.Discard
+	hook := NewHook(token, env)
```
Comment (tamccall) [inline benchmark_test.go:34]: May want to explicitly set the buffer here similar to line 53 to make the benchmark more understandable

File: rollrus.go [hunk 4]
```diff
@@ -2,60 +2,120 @@ package rollrus
...
+	LogLevels  []log.Level
+}
+
 var defaultTriggerLevels = []log.Level{
 	log.ErrorLevel,
 	log.FatalLevel,
 	log.PanicLevel,
 }
 
+var defaultNumWorkers = 8 * runtime.NumCPU()
```
Comment (tamccall) [inline rollrus.go:35]: Could be lower. I'll see what everyone else thinks

File: .travis.yml [hunk 5]
```diff
@@ -1,6 +1,7 @@
 language: go
 go:
-- 1.6.3
```
Comment (greggjs) [inline .travis.yml]: If we actually wanted to get this merged in to rollrus, would we need to support this?

File: buffer/channel/buffer.go [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+	c.value = next
+	return ok
+}
+
+func (c *Buffer) Value() *logrus.Entry {
+	return c.value
+}
+
+func (c *Buffer) Push(entry *logrus.Entry) {
+	if c.closed {
```
Comment (greggjs) [inline buffer/channel/buffer.go]: why not just do this instead
```go
if !c.closed {
         c.c <- entry
}
```

**Review by greggjs — APPROVED** [review]

File: .travis.yml (hunk 5 above)
Comment (tamccall) [reply .travis.yml]: probably. It would be up to them. 1.6 is getting pretty old

File: buffer/channel/buffer.go (hunk 6 above)
Comment (tamccall) [reply buffer/channel/buffer.go]: personal preference. I'll update

File: worker.go [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+	"github.com/stvp/roll"
+)
+
+type job struct {
+	client roll.Client
+	entry  *log.Entry
+}
+
+func (j job) sendToRollbar() {
+	entry := j.entry
```
Comment (sephnewman) [inline worker.go:19]: should make a check for `if entry == nil`

Comment (tamccall) [reply worker.go:19]: 👍

**Review by greggjs — APPROVED** [review]
LGTM 👍

**Pull request merged**